source/s3_client.c

```diff
@@ -0,0 +1,168 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include "s3_client.h"
+
+#include "auth.h"
+#include "io.h"
+
+static const char *s_capsule_name_s3_client = "aws_s3_client";
+
+struct s3_client_binding {
+    struct aws_s3_client *native;
+
+    bool release_called;
+    bool shutdown_called;
+
+    /* Shutdown callback, reference cleared after setting result */
+    PyObject *on_shutdown;
+};
+
+static void s_s3_client_release(struct s3_client_binding *client) {
+    AWS_FATAL_ASSERT(!client->release_called);
+    client->release_called = true;
+
+    bool destroy_after_release = client->shutdown_called;
+
+    aws_s3_client_release(client->native);
+
+    if (destroy_after_release) {
+        aws_mem_release(aws_py_get_allocator(), client);
+    }
+}
+
+/* Invoked when the python object get cleaned up */
+static void s_s3_client_capsule_destructor(PyObject *capsule) {
+    struct s3_client_binding *client = PyCapsule_GetPointer(capsule, s_capsule_name_s3_client);
+    s_s3_client_release(client);
+}
+
+/* Callback from C land, invoked when the underlying shutdown process finished */
+static void s_s3_client_shutdown(void *user_data) {
+    struct s3_client_binding *client = user_data;
+
+    /* Lock for python */
+    PyGILState_STATE state;
+    if (aws_py_gilstate_ensure(&state)) {
+        return; /* Python has shut down. Nothing matters anymore, but don't crash */
+    }
+
+    client->shutdown_called = true;
+
+    bool destroy_after_shutdown = client->release_called;
+
+    /* Invoke on_shutdown, then clear our reference to it */
+    PyObject *result = PyObject_CallFunction(client->on_shutdown, NULL);
+    if (result) {
+        Py_DECREF(result);
+    } else {
+        /* Callback might fail during application shutdown */
+        PyErr_WriteUnraisable(PyErr_Occurred());
+    }
+    Py_CLEAR(client->on_shutdown);
+
+    if (destroy_after_shutdown) {
```

Comment:
I think you need to move the
```
client->shutdown_called = true;
bool destroy_after_shutdown = client->release_called;
```
lines down here, after you've called `Py_CLEAR(client->on_shutdown)`

Imagine this scenario:
- python object is still alive
- s_s3_client_shutdown() runs
- sets shutdown_called = true
- invokes on_shutdown
  - from on_shutdown callback, user drops reference to python object
  - now capsule destructor runs, s_s3_client_release runs, sees that shutdown_called==true and release_called==true, so destroys memory
- on_shutdown callback returns, you run `Py_CLEAR(client->on_shutdown);` but crash because client memory has been deleted

I fixed this exact same kind of bug a month ago so it's fresh in my mind